ee/session_recordings/ai/embeddings_runner.py

```diff
@@ -0,0 +1,278 @@
+import json
+import tiktoken
+import datetime
+import pytz
+
+from typing import Dict, Any, List
+
+from abc import ABC, abstractmethod
+from prometheus_client import Histogram, Counter
+from structlog import get_logger
+from openai import OpenAI
+
+from posthog.models import Team
+from posthog.clickhouse.client import sync_execute
+
+from posthog.session_recordings.queries.session_replay_events import SessionReplayEvents
+from ee.session_recordings.ai.utils import (
+    SessionSummaryPromptData,
+    reduce_elements_chain,
+    simplify_window_id,
+    format_dates,
+    collapse_sequence_of_events,
+    only_pageview_urls,
+)
+
+# tiktoken.encoding_for_model(model_name) specifies encoder
+# model_name = "text-embedding-3-small" for this usecase
+encoding = tiktoken.get_encoding("cl100k_base")
+
+MAX_TOKENS_FOR_MODEL = 8191
+
+RECORDING_EMBEDDING_TOKEN_COUNT = Histogram(
+    "posthog_session_recordings_recording_embedding_token_count",
+    "Token count for individual recordings generated during embedding",
+    buckets=[0, 100, 500, 1000, 2000, 3000, 4000, 5000, 6000, 8000, 10000],
+)
+
+GENERATE_RECORDING_EMBEDDING_TIMING = Histogram(
+    "posthog_session_recordings_generate_recording_embedding",
+    "Time spent generating recording embeddings for a single session",
+    buckets=[0.1, 0.2, 0.3, 0.4, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18, 20],
+)
+
+SESSION_EMBEDDINGS_GENERATED = Counter(
+    "posthog_session_recordings_embeddings_generated",
+    "Number of session embeddings generated",
+)
+
+SESSION_EMBEDDINGS_FAILED = Counter(
+    "posthog_session_recordings_embeddings_failed",
+    "Instance of an embedding request to open AI (and its surrounding work) failing and being swallowed",
+)
+
+SESSION_EMBEDDINGS_FATAL_FAILED = Counter(
+    "posthog_session_recordings_embeddings_fatal_failed",
+    "Instance of the embeddings task failing and raising an exception",
+)
+
+SESSION_EMBEDDINGS_WRITTEN_TO_CLICKHOUSE = Counter(
+    "posthog_session_recordings_embeddings_written_to_clickhouse",
+    "Number of session embeddings written to Clickhouse",
+)
+
+SESSION_SKIPPED_WHEN_GENERATING_EMBEDDINGS = Counter(
+    "posthog_session_recordings_skipped_when_generating_embeddings",
+    "Number of sessions skipped when generating embeddings",
+)
+
+SESSION_EMBEDDINGS_FAILED_TO_CLICKHOUSE = Counter(
+    "posthog_session_recordings_embeddings_failed_to_clickhouse",
+    "Number of session embeddings failed to Clickhouse",
+)
+
+
+logger = get_logger(__name__)
+
+
+class EmbeddingsRunner(ABC):
+    team: Team
+    source_type: str
+    client: Any
+
+    def __init__(
+        self,
+        team: Team,
+    ):
+        self.team = team
+        self.client = OpenAI()
```

Comment:
Since we're only varying preparation. I'd be tempted to pass in an `EmbeddingPreparation` over inheriting from a base

We should then also pass in a `label_name` and have that on logs and as labels on the prometheus counters so that we can break graphs up by flow...

e.g. is the error flow much faster or slower? does it have different token constraints, does it fail more or less?